feat(lint): implement vue/noSetupPropsReactivityLoss

AsherDe opened the pull request:
## Summary

This PR introduces a new lint rule, `vue/noSetupPropsReactivityLoss`, ported from `eslint-plugin-vue` to prevent a common bug in Vue's Composition API.

The **motivation** for this change is to help developers avoid accidentally losing props reactivity. In Vue, destructuring the `props` object directly in the `setup` function's parameters (e.g., `setup({ count })`) or in its root scope (e.g., `const { count } = props`) breaks the reactive link. This can lead to components that do not update correctly when their props change, which can be difficult to debug.

This new rule identifies these specific patterns and provides a clear diagnostic to the user. It correctly handles common exceptions, ignoring destructuring that occurs within nested functions (like `watch` callbacks or returned render functions) and when using reactivity-preserving helpers like `toRefs`.

A changeset has been added for this new feature.

Fixes #6305

## Test Plan

The rule's implementation is verified through a comprehensive suite of spec (snapshot) tests to ensure its correctness and robustness across various edge cases.

* **Invalid cases** that correctly trigger the rule have been tested, including:
    * Simple destructuring in `setup` function parameters (`setup({ count })`).
    * Simple destructuring via variable assignment (`const { count } = props`).
    * Destructuring of multiple properties, with renaming, and with default values.
    * Usage of `const`, `let`, and `var` for assignments.
    * Nested object destructuring and array destructuring.

* **Valid cases** that are correctly ignored by the rule have also been thoroughly tested:
    * Accessing props via direct member expression (`props.count`).
    * Destructuring that occurs inside nested functions (e.g., within `watch` callbacks or returned render functions).
    * Using reactivity-preserving helpers like `toRefs()`.
    * Destructuring from a local variable that is not the `props` object.
    * Different `setup` function signatures (e.g., with a `context` parameter, as an arrow function, or with no parameters).
    * Components that do not have a `setup` method at all.

## Docs

As this is a new lint rule, the documentation is included directly within the rule's source code (`declare_lint_rule!`) as `rustdoc` comments. This documentation explains the purpose of the rule and provides `Invalid` and `Valid` code examples, which are also used in the spec tests.

No separate website PR is required for the rule's core documentation at this time.

ematipico commented on .changeset/yellow-parts-appear.md:
```diff
@@ -0,0 +1,18 @@
+"@biomejs/biome": minor
+---
+
+Added the `vue/noSetupPropsReactivityLoss` rule to the nursery.
```
The name of the rule is wrong.

ematipico replied:
```suggestion
Added the nursery rule`vue/noSetupPropsReactivityLoss`.
```

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+use biome_rule_options::no_setup_props_reactivity_loss::NoSetupPropsReactivityLossOptions;
+
+declare_lint_rule! {
+    /// Disallow usages that lose the reactivity of `props` passed to `setup`.
```
```suggestion
    /// Disallow usages that lose the reactivity of `props` passed to `setup` in Vue projects. 
```

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+    ///
+    pub NoSetupPropsReactivityLoss {
+        version: "next",
+        name: "noSetupPropsReactivityLoss",
```
Please refer to the contribution guide for the correct name of rule https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md#naming-conventions-for-rules

dyc3 replied:
This is my mistake, in the linked issue the name `noSetupPropsReactivityLoss` is exactly what I suggested, and I forgot to updated it to reflect current guidelines.

The rule name should be `noVueSetupPropsReactivityLoss`.

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
````diff
@@ -0,0 +1,386 @@
+    /// }
+    /// ```
+    ///
+    pub NoSetupPropsReactivityLoss {
````
This rule must be added to the vue domain, and we must add the source too

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+    type Signals = Vec<Self::State>;
+    type Options = NoSetupPropsReactivityLossOptions;
+
+    /// Analyzes Vue component exports to detect props reactivity loss patterns.
+    ///
+    /// This method examines the default export of a Vue component, specifically
+    /// looking for setup methods that might cause props to lose reactivity through
+    /// destructuring or improper usage patterns.
```
Please move this comment to line 112

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+        let export_clause = ctx.query();
+        let mut signals = Vec::new();
+
+        if let Ok(expression) = export_clause.expression()
+            && let Some(obj_expr) = expression.as_js_object_expression()
+            && let Some(setup_method) = find_setup_method(obj_expr)
```
There's no need to use let chains. You can use the try operator, for example 

```rs
let expression = export_clause.expression().ok()?;
```

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+        signals
+    }
+
+    /// Creates diagnostic messages for detected reactivity loss violations.
+    ///
+    /// Generates appropriate error messages and helpful notes based on the type
+    /// of violation detected (parameter destructuring vs assignment destructuring).
```
Please remove this comment

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+/// Searches for a 'setup' method within a Vue component object expression.
+///
+/// Iterates through the object members to find a method with the name 'setup',
+/// which is the Composition API setup function in Vue components.
```
```suggestion
/// which is the Composition API setup function in Vue component object expressions.
```

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+        let name = method.name().ok()?;
+        let literal_name = name.as_js_literal_member_name()?;
+        let name_token = literal_name.name().ok()?;
+        (name_token.text() == "setup").then(|| method.clone())
```
```suggestion
        (name_token.text_trimmed() == "setup").then(|| method.clone())
```

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+}
+
+impl Rule for NoSetupPropsReactivityLoss {
+    type Query = Semantic<JsExportDefaultExpressionClause>;
```
This rule isn't using the semantic model. Replace `Semantic` with the `Ast` query

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+fn check_setup_body_statements(
+    statements: impl Iterator<Item = AnyJsStatement>,
+    signals: &mut Vec<RuleState>,
+    params: &Option<biome_js_syntax::JsParameters>,
```
It's more idiomatic to send `Option<&>`. You can use `as_ref()` to achieve that

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+                .name_token()
+                .ok()
+        })
+        .map(|n| n.text().to_string());
```
Please don't allocate a string. Use `to_trimmed_text` instead

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+fn check_variable_declaration_for_props_destructuring(
+    declaration: &JsVariableDeclaration,
+    signals: &mut Vec<RuleState>,
+    props_param_name: &Option<String>,
```
In Rust it's not idiomatic passing this. You should pass `Option<&str>`.

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+fn check_assignment_expression_for_props_destructuring(
+    assign_expr: &JsAssignmentExpression,
+    signals: &mut Vec<RuleState>,
+    props_param_name: &Option<String>,
```
Ditto

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+    /// of violation detected (parameter destructuring vs assignment destructuring).
+    fn diagnostic(_ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        match state {
+            RuleState::DestructuredParameter(range) => Some(
+                RuleDiagnostic::new(
+                    rule_category!(),
+                    *range,
+                    markup! {
+                        "Destructuring props in setup function parameters causes reactivity loss."
+                    },
+                )
+                .note(markup! {
+                    "Use the props object directly to maintain reactivity: `props.propertyName`"
```
This diagnostic doesn't contain the first pillar. We need to message error first, then explain why 

https://biomejs.dev/linter/#rule-pillars

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+    /// - Destructuring assignment of props in the root scope of setup
+    /// - Member access patterns that could break reactivity
+    ///
+    /// Note: Destructuring is allowed inside nested functions, callbacks, and
```
```suggestion
    /// Note: destructuring is allowed inside nested functions, callbacks, and
```

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,386 @@
+    /// in the root scope of `setup` will cause the values to lose reactivity.
+    ///
+    /// This rule reports:
+    /// - Direct destructuring of props in setup function parameters
+    /// - Destructuring assignment of props in the root scope of setup
+    /// - Member access patterns that could break reactivity
```
```suggestion
    /// - direct destructuring of props in setup function parameters.
    /// - Destructuring assignment of props in the root scope of setup
    /// - Member access patterns that could break reactivity.
```

ematipico reviewed (changes requested):
Thank you @AsherDe for your contribution. There are some things that must be addressed 

If you used some used AI tools for this PR, please disclose it.

ematipico commented:
By the way, you might have noticed that using a "top to bottom" strategy might be very challenging, because you need to track all places where props are used. The current implementation doesn't cut it.

You two options:
- use a bottom to up strategy, similar to what we do with a similar rule https://github.com/biomejs/biome/blob/main/crates%2Fbiome_js_analyze%2Fsrc%2Flint%2Fcorrectness%2Fno_solid_destructured_props.rs
- use the semantic model to track where the props are read

AsherDe commented:
@ematipico Thanks so much for this detailed and insightful feedback.

To be honest, I was actually struggling with the top-down approach, wondering how I could reliably cover all the different ways `props` might be passed around and used. Your explanation makes perfect sense and clarifies the exact problem I was wrestling with.

I will go with **Option 1** and refactor the rule using the bottom-up strategy. The `no_solid_destructured_props.rs` example is a fantastic guide, and I'm confident I can rework the logic based on that pattern.

Thanks again for the great guidance! I'll push the updated implementation once it's ready.

AsherDe commented:
There's a known issue with the documentation validation CI check failing. However, the rule itself works perfectly— all 8 snapshot tests pass. The failure appears to be due to a technical limitation in the doc generator's parsing environment and does not affect the rule's functionality.

AsherDe commented:
Hi @ematipico ,

Thanks again for the great feedback and guidance. Your suggestion to switch to a bottom-up strategy was spot on, and your `no_solid_destructured_props.rs` rule was the perfect reference to guide the refactor.

I've pushed the new implementation. Here’s a quick overview of the changes:

* **New Strategy**: The rule now uses a "bottom-up" approach. The `Query` entry point is now `AnyJsFunction` instead of the component export.

* **Identifying `setup`**: I've added a new helper function, `is_vue_setup_function`, which is responsible for determining if a given function is actually a `setup` method within a valid Vue component (it checks for `export default`, `defineComponent`, etc.).

* **Detecting Violations**:
    * For **parameter destructuring**, the logic remains straightforward, checking the function's first parameter.
    * For **root-scope destructuring**, the rule now uses the `SemanticModel`. It finds the binding for the `props` parameter and then iterates through all its reads. For each read, it verifies two things:
        1.  Is the reference directly within the root scope of the `setup` function?
        2.  Is it part of a destructuring assignment?

* **Other Fixes**: I also incorporated the other review feedback, so the rule now correctly handles namespaced reactive helpers (e.g., `Vue.toRefs`) and quoted property names (e.g., `"setup": ...`).

I believe this new implementation is much more robust and correctly handles the edge cases. It's ready for another look whenever you have a chance. Thanks again for your help!

AsherDe commented on crates/biome_js_analyze/tests/quick_test.rs:
My apologies. I have reverted the file.

AsherDe commented on crates/biome_js_analyze/tests/specs/nursery/noVueSetupPropsReactivityLoss/invalid-conditional-destructuring.js.snap:
Addressed. I've gone through each one to confirm they are all correct now.

AsherDe commented on crates/biome_js_analyze/src/lint/nursery/no_vue_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,581 @@
+    violations
+}
+
+fn get_object_member_name_text(name: &AnyJsObjectMemberName) -> Option<String> {
```
Thanks, I've updated the code to use TokenText now, avoiding unnecessary string allocations.

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_vue_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,775 @@
+        biome_js_syntax::AnyJsObjectMemberName::JsMetavariable(_) => false,
+    }
+}
+fn get_function_first_parameter(func: &SetupFunction) -> Option<AnyJsBindingPattern> {
```
The function returns a binding, not a pattern. You might want to document the function, because the same isn't descriptive enough

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_vue_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,775 @@
+    }
+}
+
+fn get_any_js_function_first_parameter(func: &AnyJsFunction) -> Option<AnyJsBindingPattern> {
```
Same here. Let's document the business logic

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_vue_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,775 @@
+
+    for reference in model.as_binding(props_binding).all_reads() {
+        if let Some(identifier) =
+            biome_js_syntax::JsReferenceIdentifier::cast(reference.syntax().clone())
```
Remove the name of the crate, import the type at the beginning of the file.

Also, use `cast_ref` instead. Need to use `clone`

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_vue_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,775 @@
+    violations
+}
+
+fn is_reference_in_root_scope_of_function(
```
Please document the function

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_vue_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,775 @@
+    nearest_function.is_some_and(|func_node| func_node == *function_syntax)
+}
+
+fn is_function_like_node(node: &biome_rowan::SyntaxNode<biome_js_syntax::JsLanguage>) -> bool {
+    use biome_js_syntax::JsSyntaxKind;
+    matches!(
+        node.kind(),
+        JsSyntaxKind::JS_FUNCTION_DECLARATION
+            | JsSyntaxKind::JS_FUNCTION_EXPRESSION
+            | JsSyntaxKind::JS_ARROW_FUNCTION_EXPRESSION
+            | JsSyntaxKind::JS_METHOD_CLASS_MEMBER
+            | JsSyntaxKind::JS_METHOD_OBJECT_MEMBER
+            | JsSyntaxKind::JS_GETTER_CLASS_MEMBER
+            | JsSyntaxKind::JS_GETTER_OBJECT_MEMBER
+            | JsSyntaxKind::JS_SETTER_CLASS_MEMBER
+            | JsSyntaxKind::JS_SETTER_OBJECT_MEMBER
+    )
+}
```
I believe we already have a node for this, but I might be wrong. Check for `AnyFunctionLike` or something similar

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_vue_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,775 @@
+
+    for ancestor in current.ancestors() {
+        if let Some(declarator) = biome_js_syntax::JsVariableDeclarator::cast(ancestor.clone()) {
+            let Ok(id) = declarator.id() else {
+                continue;
+            };
+            let has_destructuring = id.as_js_object_binding_pattern().is_some()
+                || id.as_js_array_binding_pattern().is_some();
+
+            if !has_destructuring {
+                continue;
+            }
+
```
No need to use let-else. Just use `id.ok().map(...).unwrap_or_default()`

ematipico commented on crates/biome_js_analyze/src/lint/nursery/no_vue_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,775 @@
+                continue;
+            }
+
+            let Some(init) = declarator.initializer() else {
+                continue;
+            };
+            let Ok(init_expr) = init.expression() else {
+                continue;
+            };
```
You can use if-let chains here

AsherDe commented on crates/biome_js_analyze/src/lint/nursery/no_vue_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,583 @@
+///
+/// This function is robust against aliasing (`import { toRefs as myRefs } from 'vue'`)
+/// by resolving the binding of the callee and checking its original import name.
+fn is_reactive_api_call(expr: &AnyJsExpression, model: &SemanticModel) -> bool {
+    // We only care about call expressions, e.g., `toRefs(props)`.
+    let Some(call_expr) = expr.as_js_call_expression() else {
+        return false;
+    };
+    let Ok(callee) = call_expr.callee() else {
+        return false;
+    };
+
+    // Extract the reference identifier from the callee to resolve its binding
+    let binding = match &callee {
+        AnyJsExpression::JsIdentifierExpression(ident_expr) => {
+            ident_expr.name().ok().and_then(|name| model.binding(&name))
+        }
+        AnyJsExpression::JsStaticMemberExpression(member_expr) => {
+            // For cases like `Vue.toRefs(props)`, verify it's a Vue import chain
+            if !is_vue_import_chain(&callee, model) {
+                return false;
+            }
+            // Check if the member name is "toRefs"
+            return member_expr
+                .member()
+                .ok()
+                .and_then(|member| member.as_js_name().and_then(|name| name.value_token().ok()))
+                .is_some_and(|token| token.text_trimmed() == "toRefs");
+        }
+        _ => return false,
+    };
+
+    let Some(binding) = binding else {
+        return false;
+    };
+
+    // Verify the binding is from a Vue module AND that the original imported name was "toRefs".
+    is_vue_reactive_function_binding(&binding) && is_imported_as_reactive_function(&binding)
+}
```
I accidentally deleted this case while reordering the functions. It has been fixed.

ematipico commented:
No worries @AsherDe 

BTW, you're not forced to get the rule 100% at the first PR. You're actually encouraged to descope the rule, and break it down in multiple PRs if you think it's too much work.

AsherDe commented on crates/biome_js_analyze/src/lint/nursery/no_vue_setup_props_reactivity_loss.rs:
```diff
@@ -0,0 +1,603 @@
+    }
+}
+
+fn is_member_named_setup(name: &AnyJsObjectMemberName) -> bool {
+    name.name().is_some_and(|text| text.text() == "setup")
+}
```
AnyJsObjectMemberName::name() has correctly handled JsLiteralMemberName.
